### ISSUE TITLE
`env_reward` does not match the formula in the paper!

### ISSUE THREAD

**Issue by FolSpark** [Visual-Agent/DeepEyes#83]
In your paper, you claimed that the tool - call reward is independent of the number of calls. However, after carefully examining your code, 
in `verl/workers/agent/parallel_env.py:L215`

![Image](https://github.com/user-attachments/assets/6696fcc7-d4d8-4a58-8566-10c6ea110d59)

![Image](https://github.com/user-attachments/assets/bdc6bbf6-2cc2-40d6-adc2-bdbe0e0a8f28)

and `verl/workers/reward_manager/naive.py:L51` 

![Image](https://github.com/user-attachments/assets/1479033d-beb1-4be3-b858-9e1bfe7063c8)

It shows that: env_reward is the number of calls * 0.5, and it has nothing to do with whether the answer is accurate or not.
Does this mean that your paper does not match the code? Or is there a problem with my understanding?

Looking forward to your reply!

**Comment by FolSpark** [comment]
> The same question.
> 
> And it seems that the tool will always return reward=0 actually, as implemented below:
> 
> [DeepEyes/verl/workers/agent/envs/mm_process_engine/visual_toolbox_v2.py](https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox_v2.py#L102-L109)
> 
> Lines 102 to 109 in [561293d](/Visual-Agent/DeepEyes/commit/561293def6dc71fa7ac8b5bc674c070c393c9d94)
> 
>  obs = { 
>      "prompt": "\n<|im_start|>user\n" + "<tool_response>" +"<image>" + self.user_prompt + "</tool_response>" + "<|im_end|>\n<|im_start|>assistant\n", 
>      "multi_modal_data": {"image": [current_image]} 
>  } 
>  reward = 0.0  # Reward for successful tool call with correct JSON 
>  done = False 
>  info = {"status": "success", "tool_used": tool_name} 
>  print(f'[DEBUG] SUCCESS ACTION {action_string=}') 
> I would appreciate if someone can explain how the reward is computed.

Sorry, if the return value here is always 0, then it should be fine. I've only checked[DeepEyes/verl/workers/agent/envs/mm_process_engine/visual_toolbox.py](https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox.py#L99)
The return value here is 0.5, which leads to the reward for the number of tool uses being stacked later. I didn't notice that the author actually used version v2 of the dataset before. It was my mistake. I will close this issue.

**Comment by FolSpark** [comment]
> Oh I didn’t realize this bug when I previously tried to reproduce this repository... I’d like to ask if you know how to fix the reward bug here? I suspect that the tool call count easily dropping to 0 during training might be related to this.

I don't think this can be regarded as a bug. The author applied different tool usage strategies in different versions of the tool, for example:

https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox.py#L99

and

https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox_v2.py#L106

Here, different tool rewards will be added to 
https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/reward_manager/naive.py#L51
If v2 is used, which is 0 here, it should be consistent with the formula in the paper.

Regarding reproduction, due to my limited computing resource, I have only run the 3B model, and the number of tool calls later dropped to zero. 😭
If you use v1, since it offers cumulative rewards for the number of tool uses, it should increase the number of tool calls. You can give it a try.
This is my understanding. If there are any mistakes, please feel free to discuss.

**Comment by mengzchen** [comment]
@FolSpark Hello! I'd like to discuss the reward part with you. Do you mean that if we use the v2 version of the tool, then the *tool\_reward* term in the paper's formula only depends on whether the final answer is correct? Also, I'd like to ask: does this mean that only the final round containing `<answer>` receives a reward, and all the previous `<tool_call>` rounds have a reward of 0?

**Comment by FolSpark** [comment]
> [@FolSpark](https://github.com/FolSpark) Hello! I'd like to discuss the reward part with you. Do you mean that if we use the v2 version of the tool, then the _tool_reward_ term in the paper's formula only depends on whether the final answer is correct? Also, I'd like to ask: does this mean that only the final round containing `<answer>` receives a reward, and all the previous `<tool_call>` rounds have a reward of 0?

I'm not quite sure about your question. Are you asking if only the answer part in the entire sequence will have reward calculation? If this is the question, I don't think so, because the reward he uses is the GRPO result reward. Specifically, reward_tensor is returned in reward_manager, and the same reward will be added to all parts of the response (by means of PyTorch's broadcasting mechanism).

**Comment by mengzchen** [comment]
> > [@FolSpark](https://github.com/FolSpark) Hello! I'd like to discuss the reward part with you. Do you mean that if we use the v2 version of the tool, then the _tool_reward_ term in the paper's formula only depends on whether the final answer is correct? Also, I'd like to ask: does this mean that only the final round containing `<answer>` receives a reward, and all the previous `<tool_call>` rounds have a reward of 0?
> 
> I'm not quite sure about your question. Are you asking if only the answer part in the entire sequence will have reward calculation? If this is the question, I don't think so, because the reward he uses is the GRPO result reward. Specifically, reward_tensor is returned in reward_manager, and the same reward will be added to all parts of the response (by means of PyTorch's broadcasting mechanism).

Yes, that's exactly what I was asking. But I haven't found any claim or documentation saying that GRPO can automatically broadcast the reward to all tokens in the response. Do you happen to know which part of the code in verl this corresponds to? I'd like to study it further. Thank you!

**Comment by FolSpark** [comment]
> Yes, that's exactly what I was asking. But I haven't found any claim or documentation saying that GRPO can automatically broadcast the reward to all tokens in the response. Do you happen to know which part of the code in verl this corresponds to? I'd like to study it further. Thank you!

https://github.com/volcengine/verl/blob/d9a6a31c8d92030f456816d7e6b0c56b117fb192/verl/workers/reward_manager/naive.py#L98

Assign a scalar value to a `torch.tensor` tensor

**Comment by cq-dong** [comment]
> > Oh I didn’t realize this bug when I previously tried to reproduce this repository... I’d like to ask if you know how to fix the reward bug here? I suspect that the tool call count easily dropping to 0 during training might be related to this.
> 
> I don't think this can be regarded as a bug. The author applied different tool usage strategies in different versions of the tool, for example:
> 
> [DeepEyes/verl/workers/agent/envs/mm_process_engine/visual_toolbox.py](https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox.py#L99)
> 
> Line 99 in [561293d](/Visual-Agent/DeepEyes/commit/561293def6dc71fa7ac8b5bc674c070c393c9d94)
> 
>  reward = 0.5  # Reward for successful tool call with correct JSON 
> and
> 
> [DeepEyes/verl/workers/agent/envs/mm_process_engine/visual_toolbox_v2.py](https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/agent/envs/mm_process_engine/visual_toolbox_v2.py#L106)
> 
> Line 106 in [561293d](/Visual-Agent/DeepEyes/commit/561293def6dc71fa7ac8b5bc674c070c393c9d94)
> 
>  reward = 0.0  # Reward for successful tool call with correct JSON 
> Here, different tool rewards will be added to
> 
> [DeepEyes/verl/workers/reward_manager/naive.py](https://github.com/Visual-Agent/DeepEyes/blob/561293def6dc71fa7ac8b5bc674c070c393c9d94/verl/workers/reward_manager/naive.py#L51)
> 
> Line 51 in [561293d](/Visual-Agent/DeepEyes/commit/561293def6dc71fa7ac8b5bc674c070c393c9d94)
> 
>  reward_tensor += data.batch['env_reward'] 
> 
> If v2 is used, which is 0 here, it should be consistent with the formula in the paper.
> Regarding reproduction, due to my limited computing resource, I have only run the 3B model, and the number of tool calls later dropped to zero. 😭 If you use v1, since it offers cumulative rewards for the number of tool uses, it should increase the number of tool calls. You can give it a try. This is my understanding. If there are any mistakes, please feel free to discuss.

Hello, I would like to ask how to understand the v2 dataset used by your author. I have not been able to determine which version of visual_toolbox_v the author used? I found out that it was indeed v2 through the printout. I observed actor_rollout_ref.rollout.agent.tool_name_key=env_name \
in the script, but I was still not sure, so I would like to ask if you are sure to use visual_toolbox_v2?
